presto-hive/src/main/java/com/facebook/presto/hive/HivePageSourceProvider.java

```diff
@@ -431,17 +461,24 @@ public int getIndex()
             Set<Integer> regularColumnIndices = new HashSet<>();
             ImmutableList.Builder<ColumnMapping> columnMappings = ImmutableList.builder();
             for (HiveColumnHandle column : columns) {
-                Optional<HiveType> coercionFrom = Optional.ofNullable(columnCoercions.get(column.getHiveColumnIndex()));
+                // will be present if the partition has a different schema (column type, column name) for the column
+                Optional<Column> partitionColumn = Optional.ofNullable(partitionSchemaDifference.get(column.getHiveColumnIndex()));
+                Optional<HiveType> coerceFrom = Optional.empty();
+                // we don't care if only the column name has changed
+                if (partitionColumn.isPresent() && !partitionColumn.get().getType().equals(column.getHiveType())) {
+                    coerceFrom = Optional.of(partitionColumn.get().getType());
+                }
+
                 if (column.getColumnType() == REGULAR) {
                     checkArgument(regularColumnIndices.add(column.getHiveColumnIndex()), "duplicate hiveColumnIndex in columns list");
-                    columnMappings.add(regular(column, regularIndex, coercionFrom));
+                    columnMappings.add(regular(column, regularIndex, coerceFrom));
```

Comment:
I will let @rschlussel decide whether this variable should be named `coercionFrom` or `coerceFrom` 😃

Reply:
I didn't mean to rename it. Renamed it back.